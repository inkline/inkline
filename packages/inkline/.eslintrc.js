module.exports = {
    root: true,
    env: {
        node: true
    },
    parserOptions: {
        parser: "babel-eslint"
    },
    extends: [
        "plugin:vue/recommended"
    ],
    rules: {
        'indent': ['warn', 4],
        'no-debugger': process.env.NODE_ENV === "production" ? 2 : 0,
        'no-prototype-builtins': 'off'
    }
};
