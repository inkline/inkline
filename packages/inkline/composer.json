{
  "name": "inkline/inkline",
  "description": "Inkline is a modern UI/UX framework for Vue.js, designed for creating flawless content-rich responsive web applications",
  "type": "css-framework",
  "license": "ISC",
  "authors": [
    {
      "name": "Alex Grozav",
      "email": "alex@grozav.com"
    }
  ],
  "require": {}
}