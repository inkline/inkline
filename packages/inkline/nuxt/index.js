const { resolve } = require('path');

const defaultOptions = {
    scss: false,
    treeShaking: false
};

module.exports = function InklineNuxt(moduleOptions = {}) {
    this.nuxt.hook('build:before', () => {
        const options = {
            ...defaultOptions,
            ...this.options.inkline,
            ...moduleOptions
        };

        this.options.css = [].concat(this.options.css || []);
        if (options.scss) {
            this.options.css.unshift('@inkline/inkline/src/index.scss');
        } else {
            this.options.css.unshift('@inkline/inkline/dist/inkline.css');
        }

        this.options.build.transpile = [].concat(this.options.build.transpile || []);
        if (options.scss || options.treeShaking) {
            this.options.build.transpile.push('@inkline/inkline');
        }

        const inklineOptions = { ...options };
        delete inklineOptions.scss;
        delete inklineOptions.treeShaking;

        // Register plugin, passing options to plugin template
        this.addPlugin({
            src: resolve(__dirname, 'plugin.template.js'),
            fileName: 'inkline.js',
            options: {
                inkline: inklineOptions,
                treeShaking: options.treeShaking
            }
        });
    });
};

module.exports.meta = require('../package.json');
