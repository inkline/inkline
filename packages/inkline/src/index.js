export * from '@inkline/inkline/src/plugin';
export * from '@inkline/inkline/src/components';
