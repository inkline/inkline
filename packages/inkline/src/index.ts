export * from '@inkline/i18n';
export * from '@inkline/utils';
export * from '@inkline/validation';
export * from '@inkline/vue';

export * from './components';
