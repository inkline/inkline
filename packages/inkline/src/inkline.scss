@import 'css/themes';
@import 'css/index';
@import 'css/utilities';
