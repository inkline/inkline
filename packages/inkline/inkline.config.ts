import { defineConfig, useTheme } from '@inkline/config';
import { useComponentThemes, useElementThemes } from './src/theme';

export default defineConfig(() => {
    useTheme();
    useElementThemes();
    useComponentThemes();
});
