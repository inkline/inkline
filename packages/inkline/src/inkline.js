import '@inkline/inkline/src/inkline.scss';
import { Inkline } from '@inkline/inkline/src/plugin';
import * as components from '@inkline/inkline/src/components';

export { Inkline } from '@inkline/inkline/src/plugin';
export * from '@inkline/inkline/src/components';

const install = Inkline.install;
Inkline.install = (Vue, options = {}) => install.call(Inkline, Vue, { components, ...options });

if (typeof window !== 'undefined' && window.Vue) {
    window.Vue.use(Inkline);
}

export default Inkline;
