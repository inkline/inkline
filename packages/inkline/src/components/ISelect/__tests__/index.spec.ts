import { fireEvent, render } from '@testing-library/vue';
import { createInkline } from '@inkline/inkline/__tests__/utils';
import { ref } from 'vue';
import { ISelect, InklineKey, FormKey, keymap } from '@inkline/inkline';

describe('Components', () => {
    describe('ISelect', () => {
        const options = [
            { id: 0, label: 'Option 1', value: 1 },
            { id: 1, label: 'Option 2', value: 2 },
            { id: 2, label: 'Option 3', value: 3 }
        ];

        const props = {
            name: 'select',
            color: 'primary',
            size: 'md',
            options,
            modelValue: options[0].id
        };

        const stubs = ['i-icon'];

        it('should be named correctly', () => {
            expect(ISelect.name).toEqual('ISelect');
        });

        it('should render correctly', () => {
            const wrapper = render(ISelect, {
                props,
                global: {
                    provide: {
                        [InklineKey as symbol]: createInkline()
                    }
                }
            });

            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('props', () => {
            describe('name', () => {
                it('should have randomly generated name uid', async () => {
                    const wrapper = render(ISelect, {
                        props: {
                            color: props.color,
                            size: props.size,
                            options: props.options,
                            modelValue: props.modelValue
                        },
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });

                    expect(wrapper.container.firstChild).toHaveAttribute(
                        'id',
                        expect.stringContaining('select')
                    );
                });
            });

            describe('options', () => {
                it('should render correctly without options', () => {
                    const wrapper = render(ISelect, {
                        props: {
                            name: 'select',
                            color: 'primary',
                            size: 'md'
                        },
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });

                    expect(wrapper.html()).toMatchSnapshot();
                });
            });
        });

        describe('computed', () => {
            describe('wrapperClasses', () => {
                it('should add classes based on props', () => {
                    const wrapper = render(ISelect, {
                        props,
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });

                    expect(wrapper.container.firstChild).toHaveClass(
                        `-${props.color}`,
                        `-${props.size}`
                    );
                });
            });

            describe('tabIndex', () => {
                it('should be -1 if disabled', () => {
                    const wrapper = render(ISelect, {
                        props: {
                            disabled: true,
                            ...props
                        },
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const input = wrapper.container.querySelector('input');

                    expect(input).toHaveAttribute('tabindex', '-1');
                });

                it('should be 1 otherwise', () => {
                    const wrapper = render(ISelect, {
                        props,
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const input = wrapper.container.querySelector('input');

                    expect(input).toHaveAttribute('tabindex', '0');
                });
            });

            describe('isClearable', () => {
                it('should be clearable if value and clearable', () => {
                    const wrapper = render(ISelect, {
                        props: {
                            clearable: true,
                            ...props
                        },
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const clear = wrapper.container.querySelector('.input-clear');

                    expect(clear).toBeVisible();
                });
            });

            describe('inputPlaceholder', () => {
                it('should be default placeholder if no value', () => {
                    const placeholder = 'Placeholder';
                    const wrapper = render(ISelect, {
                        props: {
                            ...props,
                            modelValue: undefined,
                            placeholder
                        },
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const input = wrapper.container.querySelector('input');

                    expect(input).toHaveAttribute('placeholder', placeholder);
                });
            });
        });

        describe('methods', () => {
            describe('onInput()', () => {
                it('should hide the menu, and update value', async () => {
                    const wrapper = render(ISelect, {
                        props,
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const input = wrapper.container.querySelector('input');

                    await fireEvent.click(input as Element);
                    const menu = await wrapper.getByRole('listbox');
                    const optionElements = await wrapper.getAllByRole('option');
                    await fireEvent.click(optionElements[1]);

                    expect(menu).not.toBeVisible();
                    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([options[1].id]);
                });

                it('should call parent form onInput', async () => {
                    const onBlur = vi.fn();
                    const onInput = vi.fn();
                    const wrapper = render(ISelect, {
                        props,
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline(),
                                [FormKey as symbol]: {
                                    disabled: ref(false),
                                    readonly: ref(false),
                                    onBlur,
                                    onInput
                                }
                            }
                        }
                    });
                    const input = wrapper.container.querySelector('input');

                    await fireEvent.click(input as Element);
                    const optionElements = await wrapper.getAllByRole('option');
                    await fireEvent.click(optionElements[1]);

                    expect(onInput).toHaveBeenCalled();
                });
            });

            describe('onClear()', () => {
                it('should clear input value', async () => {
                    const wrapper = render(ISelect, {
                        props: {
                            ...props,
                            clearable: true
                        },
                        global: {
                            stubs,
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const clear = wrapper.container.querySelector('.input-clear');

                    await fireEvent.click(clear as Element);

                    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([null]);
                });
            });

            describe('onFocus()', () => {
                it('should not open select if there are no options', async () => {
                    const wrapper = render(ISelect, {
                        props: {
                            ...props,
                            options: []
                        },
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const input = wrapper.container.querySelector('input');
                    const menu = wrapper.container.querySelector('[role="listbox"]');

                    await fireEvent.focus(input as Element);

                    expect(menu).not.toBeVisible();
                });

                it('should open select', async () => {
                    const wrapper = render(ISelect, {
                        props,
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const input = wrapper.container.querySelector('input');

                    await fireEvent.click(input as Element);
                    const menu = await wrapper.getByRole('listbox');

                    expect(menu).toBeVisible();
                });
            });

            describe('onBlur()', () => {
                it('should hide menu, set input value and call parent.onblur', async () => {
                    const onBlur = vi.fn();
                    const onInput = vi.fn();
                    const wrapper = render(ISelect, {
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline(),
                                [FormKey as symbol]: {
                                    disabled: ref(false),
                                    readonly: ref(false),
                                    onBlur,
                                    onInput
                                }
                            }
                        },
                        props
                    });
                    const input = wrapper.container.querySelector('input') as Element;

                    await fireEvent.blur(input);

                    expect(onBlur).toHaveBeenCalled();
                });
            });

            describe('onClick()', () => {
                it('should show menu', async () => {
                    const wrapper = render(ISelect, {
                        props,
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const input = wrapper.container.querySelector('input') as Element;

                    await fireEvent.click(input);
                    const menu = await wrapper.getByRole('listbox');

                    expect(menu).toBeVisible();
                });
            });

            describe('onCaretClick()', () => {
                it('should open the input', async () => {
                    const wrapper = render(ISelect, {
                        props,
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const caret = wrapper.container.querySelector('.select-caret');

                    await fireEvent.click(caret as Element);

                    const dropdown = await wrapper.getByRole('listbox');
                    expect(dropdown).toBeVisible();
                });
            });

            describe('onTriggerKeyDown()', () => {
                it('should not do anything if triggerKeyBindings is empty', async () => {
                    const wrapper = render(ISelect, {
                        props: {
                            triggerKeyBindings: [],
                            ...props
                        },
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                    const menu = wrapper.container.querySelector('[role="listbox"]');
                    await fireEvent.keyDown(trigger, { key: keymap.up[1] });
                    expect(menu).not.toBeVisible();
                });

                ['up', 'down'].forEach((key) => {
                    describe(key, () => {
                        it('should open the dropdown', async () => {
                            const wrapper = render(ISelect, {
                                props,
                                global: {
                                    provide: {
                                        [InklineKey as symbol]: createInkline()
                                    }
                                }
                            });
                            const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                            await fireEvent.keyDown(trigger, {
                                key: keymap[key][1]
                            });
                            const menu = await wrapper.getByRole('listbox');
                            expect(menu).toBeVisible();
                        });
                    });
                });

                describe('enter', () => {
                    it('should open the dropdown', async () => {
                        const wrapper = render(ISelect, {
                            props,
                            global: {
                                provide: {
                                    [InklineKey as symbol]: createInkline()
                                }
                            }
                        });
                        const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                        await fireEvent.keyDown(trigger, {
                            key: keymap.enter[0]
                        });
                        const menu = await wrapper.getByRole('listbox');
                        expect(menu).toBeVisible();
                    });
                });

                ['tab', 'esc'].forEach((key) => {
                    describe(key, () => {
                        it('should close the dropdown', async () => {
                            const wrapper = render(ISelect, {
                                props,
                                global: {
                                    provide: {
                                        [InklineKey as symbol]: createInkline()
                                    }
                                }
                            });
                            const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                            await fireEvent.click(trigger);
                            const menu = await wrapper.getByRole('listbox');
                            await fireEvent.keyDown(trigger, {
                                key: keymap[key][0]
                            });
                            expect(menu).not.toBeVisible();
                        });
                    });
                });
            });

            describe('onItemKeyDown()', () => {
                it('should not do anything if itemKeyBindings is empty', async () => {
                    const wrapper = render(ISelect, {
                        props: {
                            itemKeyBindings: [],
                            ...props
                        },
                        global: {
                            provide: {
                                [InklineKey as symbol]: createInkline()
                            }
                        }
                    });
                    const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                    await fireEvent.click(trigger);
                    const menuItems = await wrapper.getAllByRole('option');
                    await fireEvent.keyDown(menuItems[0], {
                        key: keymap.up[1]
                    });
                    expect(menuItems[0]).not.toHaveFocus();
                });

                describe('up', () => {
                    it('should focus previous item', async () => {
                        const wrapper = render(ISelect, {
                            props: {
                                ...props
                            },
                            global: {
                                provide: {
                                    [InklineKey as symbol]: createInkline()
                                }
                            }
                        });
                        const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                        await fireEvent.click(trigger);
                        const menuItems = await wrapper.getAllByRole('option');
                        menuItems[1].focus();
                        await fireEvent.keyDown(menuItems[1], {
                            key: keymap.up[0]
                        });
                        expect(menuItems[0]).toHaveFocus();
                    });

                    it('should remain focused if first item', async () => {
                        const wrapper = render(ISelect, {
                            props: {
                                ...props
                            },
                            global: {
                                provide: {
                                    [InklineKey as symbol]: createInkline()
                                }
                            }
                        });
                        const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                        await fireEvent.click(trigger);
                        const menuItems = await wrapper.getAllByRole('option');
                        menuItems[0].focus();
                        await fireEvent.keyDown(menuItems[0], {
                            key: keymap.up[0]
                        });
                        expect(menuItems[0]).toHaveFocus();
                    });
                });

                describe('down', () => {
                    it('should focus next item', async () => {
                        const wrapper = render(ISelect, {
                            props: {
                                ...props
                            },
                            global: {
                                provide: {
                                    [InklineKey as symbol]: createInkline()
                                }
                            }
                        });
                        const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                        await fireEvent.click(trigger);
                        const menuItems = await wrapper.getAllByRole('option');
                        menuItems[0].focus();
                        await fireEvent.keyDown(menuItems[0], {
                            key: keymap.down[0]
                        });
                        expect(menuItems[1]).toHaveFocus();
                    });

                    it('should remain focused if last item', async () => {
                        const wrapper = render(ISelect, {
                            props: {
                                ...props
                            },
                            global: {
                                provide: {
                                    [InklineKey as symbol]: createInkline()
                                }
                            }
                        });
                        const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                        await fireEvent.click(trigger);
                        const menuItems = await wrapper.getAllByRole('option');
                        menuItems[1].focus();
                        await fireEvent.keyDown(menuItems[menuItems.length - 1], {
                            key: keymap.down[0]
                        });
                        expect(menuItems[menuItems.length - 1]).toHaveFocus();
                    });
                });

                ['enter', 'space'].forEach((key) => {
                    describe(key, () => {
                        it('should click item and focus trigger', async () => {
                            const wrapper = render(ISelect, {
                                props: {
                                    ...props
                                },
                                global: {
                                    provide: {
                                        [InklineKey as symbol]: createInkline()
                                    }
                                }
                            });
                            const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                            await fireEvent.click(trigger);
                            const menuItems = await wrapper.getAllByRole('option');
                            await fireEvent.keyDown(menuItems[0], {
                                key: keymap[key][0]
                            });
                            expect(menuItems[0]).not.toHaveFocus();
                        });

                        it('should close dropdown if hideOnItemClick', async () => {
                            const wrapper = render(ISelect, {
                                props: {
                                    ...props
                                },
                                global: {
                                    provide: {
                                        [InklineKey as symbol]: createInkline()
                                    }
                                }
                            });
                            const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                            await fireEvent.click(trigger);
                            const menu = await wrapper.getByRole('listbox');
                            const menuItems = await wrapper.getAllByRole('option');
                            await fireEvent.keyDown(menuItems[0], {
                                key: keymap[key][0]
                            });
                            expect(menu).not.toBeVisible();
                        });
                    });
                });

                ['tab', 'esc'].forEach((key) => {
                    describe(key, () => {
                        it('should close dropdown if hideOnItemClick', async () => {
                            const wrapper = render(ISelect, {
                                props: {
                                    ...props
                                },
                                global: {
                                    provide: {
                                        [InklineKey as symbol]: createInkline()
                                    }
                                }
                            });
                            const trigger: HTMLElement = wrapper.container.querySelector('input')!;
                            await fireEvent.click(trigger);
                            const menu = await wrapper.getByRole('listbox');
                            const menuItems = await wrapper.getAllByRole('option');
                            await fireEvent.keyDown(menuItems[0], {
                                key: keymap[key][0]
                            });
                            expect(menu).not.toBeVisible();
                        });
                    });
                });
            });
        });
    });
});
