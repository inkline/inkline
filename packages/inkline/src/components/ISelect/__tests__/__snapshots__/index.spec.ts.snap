// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components > ISelect > props > options > should render correctly without options 1`] = `
"<div id="select" class="select -primary -md" name="select" role="combobox" aria-haspopup="listbox" aria-owns="select-options" aria-expanded="false">
  <div class="input -primary -md -suffixed">
    <!--v-if-->
    <div class="input-field">
      <!--v-if-->
      <!--v-if--><input autocomplete="off" aria-autocomplete="none" aria-controls="select-options" placeholder="" name="select-input" type="text" tabindex="0" aria-disabled="false" readonly="" aria-readonly="true" value=""><span class="input-suffix"><!-- @slot password-toggle Slot for the password toggle button --><!--v-if--><!-- @slot clearable Slot for the clearable button --><!--v-if--><!-- @slot suffix Slot for the input suffix content --><!-- @slot suffix Slot for the select suffix content --><button class="select-caret" aria-hidden="true" role="button"></button></span>
    </div>
    <!--v-if-->
  </div>
  <transition-stub name="zoom-in-top-transition" appear="false" persisted="true" css="true">
    <div id="select-options" class="select-popup" role="listbox" aria-hidden="true" style="display: none;"><span class="arrow"></span>
      <!--v-if-->
      <div class="select-body">
        <div class="select-no-results">
          <!-- @slot no-results Slot for showing no options message --> There are no options.
        </div>
        <div class="select-options"></div>
      </div>
      <!--v-if-->
    </div>
  </transition-stub>
</div>"
`;

exports[`Components > ISelect > should render correctly 1`] = `
"<div id="select" class="select -primary -md" name="select" role="combobox" aria-haspopup="listbox" aria-owns="select-options" aria-expanded="false">
  <div class="input -primary -md -suffixed">
    <!--v-if-->
    <div class="input-field">
      <!--v-if-->
      <!--v-if--><input autocomplete="off" aria-autocomplete="none" aria-controls="select-options" placeholder="Option 1" name="select-input" type="text" tabindex="0" aria-disabled="false" readonly="" aria-readonly="true" value="Option 1"><span class="input-suffix"><!-- @slot password-toggle Slot for the password toggle button --><!--v-if--><!-- @slot clearable Slot for the clearable button --><!--v-if--><!-- @slot suffix Slot for the input suffix content --><!-- @slot suffix Slot for the select suffix content --><button class="select-caret" aria-hidden="true" role="button"></button></span>
    </div>
    <!--v-if-->
  </div>
  <transition-stub name="zoom-in-top-transition" appear="false" persisted="true" css="true">
    <div id="select-options" class="select-popup" role="listbox" aria-hidden="true" style="display: none;"><span class="arrow"></span>
      <!--v-if-->
      <div class="select-body">
        <!--v-if-->
        <div class="select-options">
          <div active="true" class="select-option -active" role="option" tabindex="0" aria-selected="true">
            <!-- @slot default Slot for select option label -->Option 1
          </div>
          <div active="false" class="select-option" role="option" tabindex="0">
            <!-- @slot default Slot for select option label -->Option 2
          </div>
          <div active="false" class="select-option" role="option" tabindex="0">
            <!-- @slot default Slot for select option label -->Option 3
          </div>
        </div>
      </div>
      <!--v-if-->
    </div>
  </transition-stub>
</div>"
`;
